### PR TITLE
feat(genesis): initialize genesis module types

### PR DESCRIPTION
**Pull request by ltacker**
Create `genesis` protobuf types
Create contructors
Create validation functions for types
Create messages
Refactor all testing utilities in a single package
Add tests on types and messages

### CODE REVIEW

File: proto/genesis/v1beta/types.proto [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    repeated google.protobuf.Any coins = 2 [(gogoproto.nullable) = false, (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.Coin", (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+}
+
+message ProposalAddValidatorPayload {
+    bytes operatorAddress = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.ValAddress"];
+    bytes consensusPubKey = 2 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.ConsAddress"];
+    google.protobuf.Any description = 3 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/x/staking/types.Description"];
+    google.protobuf.Any commissions = 4 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/x/staking/types.CommissionRates"];
+    google.protobuf.Any selfDelegation = 5 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.Coin"];
+    bytes genTx = 6;
```
Comment (ilgooz) [inline proto/genesis/v1beta/types.proto:89]: Do we run create a gentx off chain and send to SPN? @fadeev @ltacker

File: proto/genesis/v1beta/types.proto [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    string peer = 7;
```
Comment (ilgooz) [inline proto/genesis/v1beta/types.proto:90]:
```suggestion
    string peerAddress = 7;
```

Might be more explanatory. (unless it's something else?)

File: proto/genesis/v1beta/querier.proto [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	cosmos.base.query.v1beta1.PageResponse pagination = 1;
+    repeated Chain chains = 2;
+}
+
+message QueryShowChainRequest {
+	string chainID = 1;
+}
+
+message QueryShowChainResponse {
+	Chain chains = 1;
```
Comment (ilgooz) [inline proto/genesis/v1beta/querier.proto:31]:
```suggestion
	Chain chain = 1;
```

File: proto/genesis/v1beta/types.proto [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import "gogoproto/gogo.proto";
+
+message Chain {
+  string chainID = 1;
+  string creator = 2;
+  repeated string peers = 3; 
+  string sourceURL = 4; 
+  string sourceHash  = 5; 
+  int64 createdAt = 6;
+  bytes initialGenesis = 7;
```
Comment (ilgooz) [inline proto/genesis/v1beta/types.proto:16]: Should we have a `finalGenesis` as well?

**Review by ilgooz — COMMENTED** [review]
Nice work! :chains:

File: proto/genesis/v1beta/types.proto (hunk 4 above)
Comment (fadeev) [reply proto/genesis/v1beta/types.proto:16]: To me, finalizing a genesis means switching on a toggle that freezes the `CurrentGenesis`, so no changes can be applied to it.

File: proto/genesis/v1beta/types.proto (hunk 2 above)
Comment (ltacker) [reply proto/genesis/v1beta/types.proto:90]: No it is

**Comment by ltacker** [comment]
New changes:
- `AddValidatorProposal` contains, as a payload, an object of type `cosmos-sdk/types/tx.Tx` which represents the transaction generated by gentx
- The path to the change for `ChangeProposal` is in a string array

**Review by ilgooz — APPROVED** [review]
:bee:

**Review by fadeev — APPROVED** [review]
👏

**Pull request merged**